src/librustc_mir/interpret/intern.rs

```diff
@@ -0,0 +1,258 @@
+        &self.ecx
+    }
+
+    fn visit_aggregate(
+        &mut self,
+        mplace: MPlaceTy<'tcx>,
+        fields: impl Iterator<Item=EvalResult<'tcx, Self::V>>,
+    ) -> EvalResult<'tcx> {
+        if let Some(def) = mplace.layout.ty.ty_adt_def() {
+            if Some(def.did) == self.ecx.tcx.lang_items().unsafe_cell_type() {
+                // we are crossing over an `UnsafeCell`, const qualif ensures that this can only
+                // exist in the base constant, not behind references. Double checking here to make
+                // sure everything is alright
+                assert_eq!(self.mutability, Mutability::Mutable);
+            }
+        }
+        self.walk_aggregate(mplace, fields)
+    }
+
+    fn visit_primitive(&mut self, mplace: MPlaceTy<'tcx>) -> EvalResult<'tcx> {
+        // Handle Reference types, as these are the only relocations supported by const eval.
+        // Raw pointers (and boxes) are handled by the `leftover_relocations` logic.
+        let ty = mplace.layout.ty;
+        if let ty::Ref(_, _, mutability) = ty.sty {
+            match (self.mode, mutability) {
+                // all is "good and well" in the unsoundness of `static mut`
+                (InternMode::StaticMut, _) => {},
+                // immutable references are fine everywhere
+                (InternMode::StaticOrConst, hir::Mutability::MutImmutable) => {},
+                // we statically prevent `&mut T` via `const_qualif`
+                (InternMode::StaticOrConst, hir::Mutability::MutMutable) =>
+                    bug!("const qualif failed to prevent mutable references"),
+            }
+            let value = self.ecx.read_immediate(mplace.into())?;
+            // Handle trait object vtables
+            if let Ok(meta) = value.to_meta() {
+                let layout = self.ecx.layout_of(ty.builtin_deref(true).unwrap().ty)?;
+                if layout.is_unsized() {
+                    if let ty::Dynamic(..) = self.ecx.tcx.struct_tail(layout.ty).sty {
+                        if let Ok(vtable) = meta.unwrap().to_ptr() {
+                            // explitly choose `Immutable` here, since vtables are immutable, even
+                            // if the reference of the fat pointer is mutable
+                            self.intern(vtable, Mutability::Immutable)?;
+                        }
+                    }
+                }
+            }
+            let mplace = self.ecx.ref_to_mplace(value)?;
+            // Check if we have encountered this pointer+layout combination before.
+            // Only recurse for allocation-backed pointers.
+            if let Scalar::Ptr(ptr) = mplace.ptr {
+                let mutability = match mutability {
+                    hir::Mutability::MutImmutable => Mutability::Immutable,
+                    hir::Mutability::MutMutable => Mutability::Mutable,
+                };
+                match self.intern(ptr, mutability)? {
```

Comment:
Somewhere, I expected some code that "combines" the mutabilities -- like, in `& &mut T`, we are immutabily interning the `&mut T`, and thus we should proceed immutably, right? The memory behind the reference is mutable if and only if the reference itself is in a mutable context and it is an `&mut`. Is that right? It is right in the type system level, I guess the question is if we want to be bold enough to encode that here.

But even if we do not want to be that bold, there should be a comment discussing this.

Reply:
We are currently enforcing this via the `match (self.mode, mutability)` check above. But it would indeed make sense to forbid `mutability` == `Mutable` if `self.mutability` == `Immutable`. In that case we should not ICE though, but error. In the future we will probably be in a situation where users can create such code. Hmm... maybe even now, I'll do some checks to see how evil unions can be right now.

Reply:
TBH this seems like a totally reasonable combination to me, it just means we got a `& &mut`. The inner reference is still frozen though, you are not allowed to write to it.